src/typing/fields.ml

```diff
@@ -239,297 +239,266 @@ let field_access ctx mode f famode e p =
-		in
-		begin match Monomorph.classify_constraints r with
-		| CStructural(fields,is_open) ->
-			begin try
-				let f = PMap.find i fields in
-				if is_open && is_set then begin match f.cf_kind with
+				match !(a.a_status) with
+				| Statics { cl_kind = KAbstractImpl a } ->
+					type_field_by_forward_static (fun() ->
+						let mt = try module_type_of_type a.a_this with Exit -> raise Not_found in
+						let et = type_module_type ctx mt None p in
+						type_field_by_type et
+					) a
+				| _ -> raise Not_found
+			)
+		| TMono r ->
+			let mk_field () = {
+				(mk_field i (mk_mono()) p null_pos) with
+				cf_kind = Var { v_read = AccNormal; v_write = if is_set then AccNormal else AccNo }
+			} in
+			(match Monomorph.classify_constraints r with
+			| CStructural (fields,is_open) ->
+				(try
+					let f = PMap.find i fields in
+					(match f.cf_kind with
 					(* We previously inferred to read-only, but now we want to write. This can happen in cases like #8079. *)
-					| Var ({v_write = AccNo} as acc) -> f.cf_kind <- Var {acc with v_write = AccNormal}
-					| _ -> ()
-				end;
-				access f
-			with Not_found ->
-				if not is_open then
-					try
-						using_field ctx mode e i p
-					with Not_found ->
-						no_field()
-				else begin
+					| Var ({ v_write = AccNo } as acc) when is_open && is_set -> f.cf_kind <- Var { acc with v_write = AccNormal }
+					| _ -> ());
+					field_access f FHAnon
+				with Not_found when is_open ->
 					let f = mk_field() in
 					Monomorph.add_constraint r (MField f);
-					access f
-				end
-			end
-		| CTypes tl ->
-			let rec loop tl = match tl with
-				| [] ->
-					no_field()
-				| (t,_) :: tl ->
-					try
-						type_field (TypeFieldConfig.with_resume cfg) ctx {e with etype = t} i p mode with_type
-					with Not_found ->
-						loop tl
-			in
-			loop tl
-		| CUnknown ->
-			if not (List.exists (fun (m,_) -> m == r) ctx.monomorphs.perfunction) && not (ctx.untyped && ctx.com.platform = Neko) then begin
-				ctx.monomorphs.perfunction <- (r,p) :: ctx.monomorphs.perfunction;
-			end;
-			let f = mk_field() in
-			Monomorph.add_constraint r (MField f);
-			Monomorph.add_constraint r MOpenStructure;
-			access f
-		end
-	| TAbstract (a,pl) ->
-		let static_abstract_access_through_instance = ref false in
-		(try
-			let c = (match a.a_impl with None -> raise Not_found | Some c -> c) in
-			let f = PMap.find i c.cl_statics in
-			if not (has_class_field_flag f CfImpl) then begin
-				static_abstract_access_through_instance := true;
-				raise Not_found;
-			end;
-			field_access ctx mode f (FHAbstract(a,pl,c)) e p
-		with Not_found -> try
-			if does_forward a false then
-				let underlying_type = Abstract.get_underlying_type ~return_first:true a pl in
-				type_field (TypeFieldConfig.with_resume cfg) ctx {e with etype = underlying_type} i p mode with_type
-			else
+					field_access f FHAnon
+				)
+			| CTypes tl ->
+				type_field_by_list (fun (t,_) -> type_field_by_type { e with etype = t }) tl
```

Comment:
Shouldn't this fail if any one of constraints does not have specified field?

Reply:
No, this is an intersection constraint, so any type bound to the monomorph has to satisfy each constraint in that list, and thus also the one that requires the field.